### PR TITLE
SM-1238: Go Developer Instructions

### PR DESCRIPTION
**Pull request by coltonhurst**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/SM-1238

## 📔 Objective

The goal of this PR is to add instructions for third party Go developers that want to use the Go SDK in their own projects.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation
  team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed
  issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: languages/go/INSTRUCTIONS.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+Please see the main [README](./README.md) and [example](./example/example.go) for examples.
+
+## Linux
+
+1. Make sure you have the following on your system:
+    - [Go](https://go.dev/dl)
+    - A C/C++ toolchain. We recommend the [MUSL toolchain](https://musl.libc.org). You can install this on most debian based systems with: `sudo apt install musl-tools`
+2. Verify cgo is enabled and the `CC` and `CXX` flags are set with: `go env`.
+    - You can enable cgo by setting the flag: `go env -w CGO_ENABLED=1`
+    - If `CC` and `CXX` are not set, you can set them with: `go env -w CC=clang CXX=clang++`
```
Comment (tangowithfoxtrot) [inline languages/go/INSTRUCTIONS.md:16]:
```suggestion
    - If `CC` is not set, you can set it with: `go env -w CC=musl-gcc`
```

Most Linux distros don't use Clang by default. We need to use `musl-gcc` for `CC`. I don't see an equivalent `musl-g++`, but leaving `CXX` unset or the default is fine for our usage.

**Comment by tangowithfoxtrot** [comment]
Thanks! Just a minor correction for the Linux docs.

**Review by tangowithfoxtrot — CHANGES_REQUESTED** [review]

**Review by tangowithfoxtrot — DISMISSED** [review]
LGTM

File: languages/go/INSTRUCTIONS.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        - GCC is required on Windows according to [this cgo](https://go.dev/wiki/cgo) page. Microsoft recommends using the MinGW-w64 toolchain via [MSYS2](https://www.msys2.org) for up-to-date native builds of GCC [here](https://code.visualstudio.com/docs/cpp/config-mingw#_installing-the-mingww64-toolchain). MSYS2 installation instructions can be found [here](https://www.msys2.org).
+        - After following the MSYS2 installation instructions, you should be able to run the following to verify everything is installed:
+            - `gcc --version`
+            - `g++ --version`
+            - `gdb --version`
+2. Verify cgo is enabled and `CC` and `CXX` are set with: `go env`.
+    - You can enable cgo by setting the flag: `go env -w CGO_ENABLED=1`
+    - You can set `CC` and `CXX` with: `go env -w CC=gcc CXX=g++`
+3. Make sure you have added the Bitwarden Go SDK with go get: `go get github.com/bitwarden/sdk-go`
+4. Your project should build with: `go build`
```
Comment (Thomas-Avery) [inline languages/go/INSTRUCTIONS.md:46]: What about breaking the steps for each OS down into a few sections? 

I personally find this easier to read and code blocks are more convenient to copy.

Windows example below:






## Windows

### Prerequisites

- [Go](https://go.dev/dl)
- GCC

Go [documentation](https://go.dev/wiki/cgo) recommends installing the MinGW-w64 gcc compiler.

We recommend following the Visual Studio Code [guide](https://code.visualstudio.com/docs/cpp/config-mingw#_installing-the-mingww64-toolchain) to install the MinGW-w64 toolchain.

### Set Go Environment for cgo

#### Enable cgo

```shell
go env -w CGO_ENABLED=1
```

#### Set C and C++ Compilers

```shell
go env -w CC=gcc CXX=g++
```


Verify settings by running:

```shell
go env
```

### Install the Bitwarden Go SDK

Add the Bitwarden Go SDK to a project with the following command:

```shell
go get github.com/bitwarden/sdk-go
```

### Build

Validate the project builds successfully

```shell
go build
```

File: languages/go/INSTRUCTIONS.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
+# Instructions
+
+This guide is for developers who want to use the Bitwarden Go SDK module in their own Go projects.
+
+The Bitwarden SDK is written in Rust, and we utilize FFI via [cgo](https://pkg.go.dev/cmd/cgo). We currently support static linking the Bitwarden Go module on Linux (x86-64 & arm64), Mac (x86-64 & arm64), and Windows (x86-64).
```
Comment (Thomas-Avery) [inline languages/go/INSTRUCTIONS.md:5]:
```suggestion
The Bitwarden Go SDK module utilizes FFI calls to the Rust Bitwarden SDK via [cgo](https://pkg.go.dev/cmd/cgo).

The Bitwarden Go SDK module provides the following statically linked targets:

- Linux (x86-64 & arm64)
- macOS (x86-64 & arm64)
- Windows (x86-64)
```

I personally found the subject of this information unclear.
 
I think we are trying to give the reader more information on why cgo is needed?

File: languages/go/README.md [hunk 4]
```diff
@@ -10,7 +10,9 @@ managing projects and secrets, as well as a client interface to facilitate opera
 
 ## Installation
 
-Download the SDK files and place them in your Go project directory.
+You can install the [sdk-go](https://github.com/bitwarden/sdk-go) Go module like any other Go module (for example, with `go get`). You will just need to make sure you can build with [cgo](https://pkg.go.dev/cmd/cgo) on your machine.
+
+You can find instructions [here](./INSTRUCTIONS.md).
```
Comment (Thomas-Avery) [inline languages/go/README.md:15]: Should we just link to the installation instructions and any relevant information be housed there? 

I'm on the fence, so don't feel pressure to follow this suggestion.

```
Follow the installation instructions [here](./INSTRUCTIONS.md).
```

**Review by Thomas-Avery — CHANGES_REQUESTED** [review]
Looks good! Some small suggestions.

File: languages/go/README.md (hunk 4 above)
Comment (coltonhurst) [reply languages/go/README.md]: That's a good idea, it keeps all the relevant info in one place.

Fixed in https://github.com/bitwarden/sdk/pull/861/commits/39398b057aa7542f0a75efb7e6936195196fc5ae.

File: languages/go/INSTRUCTIONS.md (hunk 3 above)
Comment (coltonhurst) [reply languages/go/INSTRUCTIONS.md]: Yes, I think it's important to mention that we only support static linking and which targets we support, which brings in the question of what platforms we support. I definitely want it to be concise and clear though.

I've made some changes in https://github.com/bitwarden/sdk/pull/861/commits/730bbce055d05eeaa854d0cd405e5c1edde63953, hopefully that is more clear. Let me know what you think.

File: languages/go/INSTRUCTIONS.md (hunk 2 above)
Comment (coltonhurst) [reply languages/go/INSTRUCTIONS.md]: Good point, definitely more readable and more concise.

Updated in: https://github.com/bitwarden/sdk/pull/861/commits/df7c5cd8195b8b0f0adc717ecc978277ead5cb78.

File: languages/go/INSTRUCTIONS.md [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+- [Go](https://go.dev/dl)
+- A C toolchain
+
+[Clang](https://clang.llvm.org/get_started.html) is the default C and C++ toolchain on Mac OS. The easiest way to ensure you have the toolchain is to install the Xcode Command Line tools.
+
+You can install Clang with:
+
+
+```shell
+xcode-select -–install
```
Comment (Thomas-Avery) [inline languages/go/INSTRUCTIONS.md:72]:
```suggestion
xcode-select --install
```

Some weird formatting thing going on here. 

Copying and pasting the command fails.

File: languages/go/INSTRUCTIONS.md [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+go get github.com/bitwarden/sdk-go
+```
+
+#### Build
+
+```shell
+go build -ldflags '-linkmode external -extldflags "-static -Wl,-unresolved-symbols=ignore-all"'
+```
+
+## Mac
```
Comment (Thomas-Avery) [inline languages/go/INSTRUCTIONS.md:59]:
```suggestion
## macOS
```

**Review by Thomas-Avery — CHANGES_REQUESTED** [review]
Changes look great! 

I was able to validate on macOS and Windows. 

Two small items to look at.

File: languages/go/INSTRUCTIONS.md (hunk 6 above)
Comment (coltonhurst) [reply languages/go/INSTRUCTIONS.md]: Fixed in https://github.com/bitwarden/sdk/pull/861/commits/92b64f40c48540efda0a46405368cd5da5866c59

File: languages/go/INSTRUCTIONS.md (hunk 5 above)
Comment (coltonhurst) [reply languages/go/INSTRUCTIONS.md]: Weird... hopefully fixed in https://github.com/bitwarden/sdk/pull/861/commits/92b64f40c48540efda0a46405368cd5da5866c59

**Review by Thomas-Avery — APPROVED** [review]
Looks good, thank you!

**Review by mzieniukbw — APPROVED** [review]
LGTM

**Review by tangowithfoxtrot — APPROVED** [review]

**Pull request merged**